lib/jsdoc/util/logger.js

```diff
@@ -183,74 +191,135 @@ logger.printDebug = wrapLogFunction('debug', printToStdout);
+    this.logLevel = logLevel;
+};
+
 /**
- * Log a message at log level {@link module:jsdoc/util/logger.LEVELS.VERBOSE}.
+ * Get the current log level.
  *
- * @alias module:jsdoc/util/logger.verbose
- * @param {string} message - The message to log.
- * @param {...*=} values - The values that will replace the message's placeholders.
+ * @alias module:jsdoc/util/logger.getLevel
+ * @return {module:jsdoc/util/logger.LEVELS} The current log level.
  */
-logger.verbose = wrapLogFunction('verbose', console.info);
+Logger.prototype.getLevel = function () {
+    return this.logLevel || DEFAULT_LEVEL;
+};
+
 /**
- * Print a string at log level {@link module:jsdoc/util/logger.LEVELS.VERBOSE}. The string is not
- * terminated by a newline.
- *
- * @alias module:jsdoc/util/logger.printVerbose
- * @param {string} message - The message to log.
- * @param {...*=} values - The values that will replace the message's placeholders.
+ * Check if the log level lets us log the message.
+ * @param {module:jsdoc/util/logger.LEVELS} level
+ * @returns {boolean}
+ * @private
  */
-logger.printVerbose = wrapLogFunction('verbose', printToStdout);
+Logger.prototype._canBeLogged = function (level) {
+    return LOG_LEVELS[this.getLevel()] >= LOG_LEVELS[level];
+};
+
+
 /**
- * Log a message at log level {@link module:jsdoc/util/logger.LEVELS.WARN}.
- *
- * @alias module:jsdoc/util/logger.warn
- * @param {string} message - The message to log.
- * @param {...*=} values - The values that will replace the message's placeholders.
+ * Invoke one of the logger handlers: debug, info, fatal, warn...
+ * @private
+ * @param {module:jsdoc/util/logger.LEVELS} level
+ * @param logArgs
  */
-logger.warn = wrapLogFunction('warn', console.warn);
+Logger.prototype._invoke = function (level, logArgs) {
+    var eventName = 'logger:' + LEVELS[level].toLowerCase();
+    var handler = this._getHandler(level);
+
+    if (_.isFunction(handler) && this._canBeLogged(level)) {
+        handler.apply(this, addPrefix(logArgs, level));
+    }
+
+    logArgs.unshift(eventName);
+    this.emit.apply(this, logArgs);
+};
 
 /**
- * Set the log level.
- *
- * @alias module:jsdoc/util/logger.setLevel
- * @param {module:jsdoc/util/logger.LEVELS} level - The log level to use.
+ * Get the log handler attached to a specific log level.
+ * @private
+ * @param {module:jsdoc/util/logger.LEVELS} level
+ * @returns {Function}
  */
-logger.setLevel = function setLevel(level) {
-    logLevel = (level !== undefined) ? level : DEFAULT_LEVEL;
+Logger.prototype._getHandler = function (level) {
+    return this.handlers[level];
 };
 
+
 /**
- * Get the current log level.
- *
- * @alias module:jsdoc/util/logger.getLevel
- * @return {module:jsdoc/util/logger.LEVELS} The current log level.
+ * Attach a log handler to a specific log level.
+ * @param {module:jsdoc/util/logger.LEVELS}level
+ * @param {Function} handler
+ * @param [context]
+ */
+Logger.prototype.setHandler = function (level, handler, context) {
```

Comment:
This is probably moot, since I'm suggesting that we get rid of this method, but I'll mention it anyhow: If we were keeping this method, I'd suggest that we change the signature to `(level, handler)` and make it up to the caller to bind `handler` to a `this` object. That would also allow `setHandlers` to take a dictionary instead of an array of objects.